### PR TITLE
Make Binary cross entropy with logit numerically stable for high logit values

### PR DESCRIPTION
**Pull request by BeneSim**
See #2561 for details. I added detailed commentary to the commits.

### CODE REVIEW

**Comment by BeneSim** [comment]
Just a quick update, I also encountered the case where the logit was so small that the sigmoid returned 0. So I guess we need a better way. Maybe the method tensorflow uses might make sense, they basically calculate the log_sigmoid by using softplus:

See [log_sigmoid](https://www.tensorflow.org/api_docs/python/tf/math/log_sigmoid) and [softplus](https://www.tensorflow.org/api_docs/python/tf/math/softplus). This could be implemented like
```rust
let log_sigmoid_input = (inp.neg()?.exp()? + 1.)?.log()?.neg()?;
```

**Comment by ivarflakstad** [comment]
Noting down the pytorch [ref](https://github.com/pytorch/pytorch/blob/b5784b9e2d6cd0a7f4764e89502179245c9955b1/aten/src/ATen/native/Loss.cpp#L349) for later 👍
